model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate21_0_2_TermsAndConditionsRequiredAction.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2023 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.connections.jpa.updater.liquibase.custom;
+
+import liquibase.exception.CustomChangeException;
+import liquibase.statement.core.UpdateStatement;
+import org.keycloak.models.UserModel;
+
+public class JpaUpdate21_0_2_TermsAndConditionsRequiredAction extends CustomKeycloakTask {
+
+    private static final String TERMS_AND_CONDITION_LEGACY_ALIAS = "terms_and_conditions";
+
+    @Override
+    protected void generateStatementsImpl() throws CustomChangeException {
+        statements.add(
+            new UpdateStatement(null, null, getTableName("REQUIRED_ACTION_PROVIDER"))
+                .addNewColumnValue("ALIAS", UserModel.RequiredAction.TERMS_AND_CONDITIONS.name())
+                .addNewColumnValue("PROVIDER_ID", UserModel.RequiredAction.TERMS_AND_CONDITIONS.name())
+                .setWhereClause("ALIAS=?")
+                .addWhereParameter(TERMS_AND_CONDITION_LEGACY_ALIAS)
+        );
+
+        statements.add(
+            new UpdateStatement(null, null, getTableName("USER_REQUIRED_ACTION"))
+                .addNewColumnValue("REQUIRED_ACTION", UserModel.RequiredAction.TERMS_AND_CONDITIONS.name())
+                .setWhereClause("REQUIRED_ACTION=?")
+                .addWhereParameter(TERMS_AND_CONDITION_LEGACY_ALIAS)
```

Comment:
Some syntax comments: some quotes might be missing for the attribute name - this is probably obvious. 
Something that might be less obvious: value is a string, and it contains a number. To properly compare them, one would need to convert them to a number. I don't have access the code at the moment, but something I remember: When this is a CLOB field, different databases would require different conversions to make this a number :-( ... and this is again slow.

I see that you want to delete the older entry (smaller value), and keep the newer one. This is one way to clean it up. Still it is a choice, and it might not fit all customer situations - maybe they want to keep the other one. While this can work and is a possible solution, it should be mentioned in the migration guide. 

Some more context on T&C: Terms and conditions change, and one would have users to reconfirm the T&C now and then. So in some situations one would be interested when they confirmed them the last time. Still this depends on the use case, and the choice we're about to make should be documented in the migration guide. 

As those statements will analyze a lot of data, and possibly also update a lot of data, those statements will be slow. Nothing we can change here I'm afraid.  

So my take on this: Use the statement above, and put enough information in the migration guide (which should include the possibly long schema migration execution time).

Reply:
Looking at those who depend on the user attribute, using the lower case variant has its advantages. Having the required actions lower case or upper case is not consistent anyway.

I can see your point in going back to the old name, and I see the advantages. I don't know why the capitalization has been changed in the original PR, so my view would be limited here. Not sure if you want to have it in all places lowercase, or just for the user attribute. 

Now we have all the options on the table, I leave the decision for you.

Reply:
The capitalization was changed to be consistent with [other default-required actions](https://github.com/keycloak/keycloak/blob/main/server-spi/src/main/java/org/keycloak/models/UserModel.java#L265-L274), all of them have upper-case `providerId` and `alias`. Only T&C had it lower-case which made it inconsistent and was causing this bug: https://github.com/keycloak/keycloak/issues/9991

The idea was to revert only the user attribute. `providerId` and `alias` of T&C would still stay upper-case. In my opinion, the consistency still makes sense, especially when we already did part of it in KC21 release. 

I will update the PR to revert the user attribute and put together some text for the migration guide.